File: ui/tract-renderer/src/main/java/org/cru/godtools/tract/model/Resource.java
package org.cru.godtools.tract.model;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.WorkerThread;

import org.ccci.gto.android.common.util.XmlPullParserUtils;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;

import static org.cru.godtools.xml.Constants.XMLNS_MANIFEST;

public final class Resource extends Base {
    static final String XML_RESOURCE = "resource";
    private static final String XML_FILENAME = "filename";
    private static final String XML_SRC = "src";

    @Nullable
    private String mName;
    @Nullable
    private String mLocalName;

    @NonNull
    @WorkerThread
    static Resource fromXml(@NonNull final Manifest manifest, @NonNull final XmlPullParser parser)
            throws XmlPullParserException, IOException {
        final Resource resource = new Resource(manifest);
        resource.parse(parser);
        return resource;
    }

    private Resource(@NonNull final Manifest manifest) {
        super(manifest);
    }

    @Nullable
    public String getName() {
        return mName;
    }

    @Nullable
    public String getLocalName() {
        return mLocalName;
    }

    @WorkerThread
    private void parse(@NonNull final XmlPullParser parser) throws IOException, XmlPullParserException {
        parser.require(XmlPullParser.START_TAG, XMLNS_MANIFEST, XML_RESOURCE);

        mName = parser.getAttributeValue(null, XML_FILENAME);
        mLocalName = parser.getAttributeValue(null, XML_SRC);

        // discard any nested nodes
        XmlPullParserUtils.skipTag(parser);
    }
}

File: library/xml-model/src/main/java/org/cru/godtools/xml/model/Resource.kt
package org.cru.godtools.xml.model

import org.ccci.gto.android.common.util.xmlpull.skipTag
import org.cru.godtools.xml.XMLNS_MANIFEST
import org.xmlpull.v1.XmlPullParser

private const val XML_FILENAME = "filename"
private const val XML_SRC = "src"

class Resource : BaseModel {
    internal companion object {
        internal const val XML_RESOURCE = "resource"
    }

    internal constructor(manifest: Manifest, parser: XmlPullParser) : super(manifest) {
        parser.require(XmlPullParser.START_TAG, XMLNS_MANIFEST, XML_RESOURCE)

        name = parser.getAttributeValue(null, XML_FILENAME)
        localName = parser.getAttributeValue(null, XML_SRC)

        parser.skipTag()
    }

    val name: String?
    val localName: String?
}
